### PR TITLE
Feature keyboard event

### PR DESCRIPTION
**Pull request by saveman71**

### CODE REVIEW

File: include/lug/Window/Event.hpp [hunk 1]
```diff
@@ -1,17 +1,55 @@
 #pragma once
 
 #include <lug/Window/Export.hpp>
+#include <lug/Window/Keyboard.hpp>
+
+#include <cstdint>
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: Order of includes (or we can do the meeting of the Guidelines as soon as possible)

File: include/lug/Window/Event.hpp [hunk 2]
```diff
@@ -1,17 +1,55 @@
...
 
 namespace lug {
 namespace Window {
 
+/**
+ * Type of event
+ */
 enum class LUG_WINDOW_API EventType : uint32_t {
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: We can maybe change that and put in the `Event` structure. And after that use it with `Event::Type`. Just to preserve the coherency of the code.

File: include/lug/Window/Event.hpp [hunk 3]
```diff
@@ -1,17 +1,55 @@
...
-    CLOSE,
-    DESTROY,
+    Close,          ///< Window close event
```
Comment (Quentin01) [inline include/lug/Window/Event.hpp]: Why do you use this kind of comments, starting with `///<`?

File: include/lug/Window/Keyboard.hpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include <lug/Window/Export.hpp>
+#ifdef LUG_SYSTEM_WINDOWS
```
Comment (Quentin01) [inline include/lug/Window/Keyboard.hpp]: To preserve the coherency of the code, use `#if defined(...)`

File: include/lug/Window/Keyboard.hpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+#include <Windows.h>
+#endif
+#ifdef LUG_SYSTEM_LINUX
```
Comment (Quentin01) [inline include/lug/Window/Keyboard.hpp]: I prefer to have an `#elif` and a final `#else` with a clear error message

File: samples/window/main.cpp [hunk 6]
```diff
@@ -7,18 +9,164 @@
     #include <lug/System/Logger/OstreamHandler.hpp>
 #endif
 
+#include <unordered_map>
```
Comment (Quentin01) [inline samples/window/main.cpp]: Double include of `unordered_map`

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 7]
```diff
@@ -49,15 +52,153 @@ void lug::Window::priv::WindowImpl::close() {
...
+        case XK_ugrave:             return Keyboard::Key::Ugrave;           // The ù key
+        case XK_asterisk:           return Keyboard::Key::Asterisk;         // The * key
+        case XK_dollar:             return Keyboard::Key::Dollar;           // The $ key
+        case XK_colon:              return Keyboard::Key::Colon;            // The : key
+        case XK_exclam:             return Keyboard::Key::Exclam;           // The ! key
+        case XK_less:               return Keyboard::Key::Less;             // The < key
+        case XK_greater:            return Keyboard::Key::Greater;          // The > key
+
+        // Note: the keys for the keypad below are in the numlock off state, i.e. 8 => KP_Up
+        // It would be nice to actually find a good solution to this problem.
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Maybe add a TODO

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 8]
```diff
@@ -100,12 +257,10 @@ void lug::Window::priv::WindowImpl::setWindowDecorations(Style style) {
...
-                    PropModeReplace,
-                    reinterpret_cast<const unsigned char*>(&hints),
-                    5);
+    XChangeProperty(_display, _window, _wmHints, _wmHints,
+                    32, PropModeReplace, reinterpret_cast<const unsigned char*>(&hints), 5);
 }
+
+} // lug
+} // Window
+
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Empty line at the end of the file

File: src/lug/Window/Win32/WindowImplWin32.cpp [hunk 9]
```diff
@@ -102,7 +108,125 @@ void lug::Window::priv::WindowImpl::close() {
     ReleaseCapture();
 }
 
-bool lug::Window::priv::WindowImpl::pollEvent(lug::Window::Event& event) {
+Keyboard::Key vkToLugKey(WPARAM key) {
```
Comment (Quentin01) [inline src/lug/Window/Win32/WindowImplWin32.cpp]: `static` function? Do you want to access it outside of this file?

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 10]
```diff
@@ -49,15 +52,153 @@ void lug::Window::priv::WindowImpl::close() {
     }
 }
 
-Bool selectEvents(Display*, XEvent* event, XPointer) {
-    if (event->type == ClientMessage || event->type == DestroyNotify) {
-        return True;
-    } else {
-        return False;
+Keyboard::Key keysymToLugKey(KeySym key) {
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: `static` function? Do you want to access it outside of this file?

File: src/lug/Window/Window.cpp [hunk 11]
```diff
@@ -2,37 +2,43 @@
...
 
-    #include <lug/Window/Unix/WindowImplX11.hpp>
+#include <lug/Window/Unix/WindowImplX11.hpp>
 
 #elif defined(LUG_SYSTEM_ANDROID)
 
-    #include <lug/Window/Android/WindowImplAndroid.hpp>
+#include <lug/Window/Android/WindowImplAndroid.hpp>
 
 #endif
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: Don't we need a final `#else` with an explicit message?

File: src/lug/Window/Window.cpp [hunk 12]
```diff
@@ -55,28 +61,178 @@ bool lug::Window::Window::createWindow(uint16_t width, uint16_t height, const st
...
+        // If we've receive Destroy just close window without returning the event to the user
+        if (value && event.type == EventType::Destroy) {
             close();
             return false;
         }
+
+        if (event.type == EventType::KeyPressed) {
+            _keyState[event.key.code] = true;
+        }
+        else if (event.type == EventType::KeyReleased) {
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: `else if` at the end of the previous line

File: src/lug/Window/Window.cpp [hunk 13]
```diff
@@ -55,28 +61,178 @@ bool lug::Window::Window::createWindow(uint16_t width, uint16_t height, const st
...
     }
 }
+
+bool Window::isKeyPressed(Keyboard::Key key) const {
+    return (_keyState.at(key));
+}
+
+} // namespace Window
+} // namespace lug
+
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: Empty line at the end of the file

**Review by Quentin01 — CHANGES_REQUESTED** [review]

File: src/lug/Window/Unix/WindowImplX11.cpp [hunk 14]
```diff
@@ -1,18 +1,27 @@
+#include <lug/System/Logger.hpp>
```
Comment (Quentin01) [inline src/lug/Window/Unix/WindowImplX11.cpp]: Order of include

File: src/lug/Window/Win32/WindowImplWin32.cpp [hunk 15]
```diff
@@ -121,16 +245,47 @@ bool lug::Window::priv::WindowImpl::pollEvent(lug::Window::Event& event) {
     return false;
 }
 
-void lug::Window::priv::WindowImpl::processWindowEvents(UINT message, WPARAM /*wParam*/, LPARAM /*lParam*/) {
-    lug::Window::Event e;
+void priv::WindowImpl::enableKeyRepeat(bool enable)
+{
```
Comment (Quentin01) [inline src/lug/Window/Win32/WindowImplWin32.cpp]: Bracket at the end of the previous line

File: src/lug/Window/Window.cpp [hunk 16]
```diff
@@ -2,37 +2,49 @@
...
-lug::Window::Window::Window() : _impl{new lug::Window::priv::WindowImpl(this)} {
+namespace lug {
+namespace Window {
+
+Window::Window() : _impl{ new priv::WindowImpl(this) } {
+    initKeyState();
 }
 
-lug::Window::Window::Window(Window&& rhs) : _impl{rhs._impl} {
+Window::Window(Window&& rhs) : _impl{rhs._impl} {
```
Comment (Quentin01) [inline src/lug/Window/Window.cpp]: If think that we can't delete movable operators from `Window` since we handle them exclusively with `std::unique_ptr`

**Review by Quentin01 — CHANGES_REQUESTED** [review]

**Review by Quentin01 — APPROVED** [review]

File: samples/window/main.cpp [hunk 17]
```diff
@@ -1,4 +1,6 @@
...
 #include <lug/Window/Window.hpp>
+#include <lug/Window/Keyboard.hpp>
```
Comment (Nokitoo) [inline samples/window/main.cpp]: Order of includes

File: src/lug/Window/CMakeLists.txt [hunk 18]
```diff
@@ -13,6 +13,7 @@ set(INC
     ${INCROOT}/Export.hpp
     ${INCROOT}/Window.hpp
     ${INCROOT}/WindowHandle.hpp
+    ${INCROOT}/Keyboard.hpp
```
Comment (Nokitoo) [inline src/lug/Window/CMakeLists.txt]: Order of include files

File: src/lug/Window/Win32/WindowImplWin32.cpp [hunk 19]
```diff
@@ -121,16 +245,46 @@ bool lug::Window::priv::WindowImpl::pollEvent(lug::Window::Event& event) {
...
+        break;
+
+    case WM_KEYUP:
+    case WM_SYSKEYUP:
+        e.type = Event::Type::KeyReleased;
+        configKeyEvent(e.key, wParam, lParam);
+        break;
+
+    case WM_CHAR:
+ //   case WM_SYSCHAR:
```
Comment (Nokitoo) [inline src/lug/Window/Win32/WindowImplWin32.cpp]: Comment not indented correctly + should be written `Case`

File: src/lug/Window/Win32/WindowImplWin32.cpp [hunk 20]
```diff
@@ -193,8 +347,39 @@ bool lug::Window::priv::WindowImpl::activateFullscreen() {
...
+        new_vk = MapVirtualKey(scancode, MAPVK_VSC_TO_VK_EX);
+        break;
+    case VK_CONTROL:
+        new_vk = extended ? VK_RCONTROL : VK_LCONTROL;
+        break;
+    case VK_MENU:
+        new_vk = extended ? VK_RMENU : VK_LMENU;
+        break;
+    default:
+        // not a key we map from generic to left/right specialized
```
Comment (Nokitoo) [inline src/lug/Window/Win32/WindowImplWin32.cpp]: Missing uppercase to first letter of comment

**Pull request merged**